Mathlib/RingTheory/LocalRing/Module.lean

```diff
@@ -0,0 +1,270 @@
+    simp only [FiniteDimensional.finrank_tensorProduct, FiniteDimensional.finrank_self,
+      FiniteDimensional.finrank_finsupp_self, one_mul, add_zero]
+    rw [FiniteDimensional.finrank_eq_card_chooseBasisIndex]
+  -- On the other hand, `m ⊗ M → M` injective => `Tor₁(k, M) = 0` => `k ⊗ ker(i) → kᴵ` injective.
+  have := @lTensor_injective_of_exact_of_exact_of_rTensor_injective
+    (N₁ := LinearMap.ker i) (N₂ := I →₀ R) (N₃ := M)
+    (f₁ := (𝔪).subtype) (f₂ := Submodule.mkQ 𝔪) inferInstance inferInstance inferInstance
+    inferInstance inferInstance inferInstance
+  intro x
+  apply @this (LinearMap.ker i).subtype i (LinearMap.exact_subtype_mkQ 𝔪)
+    (Submodule.mkQ_surjective _) (LinearMap.exact_subtype_ker_map i) hi H
+    (Module.Flat.lTensor_preserves_injective_linearMap _ Subtype.val_injective)
+  apply hi'.injective
+  rw [LinearMap.baseChange_eq_ltensor]
+  erw [← LinearMap.comp_apply (i.lTensor k), ← LinearMap.lTensor_comp]
+  rw [(LinearMap.exact_subtype_ker_map i).linearMap_comp_eq_zero]
+  simp only [LinearMap.lTensor_zero, LinearMap.zero_apply, map_zero]
+
+theorem free_of_flat_of_localRing [Module.FinitePresentation R P] [Module.Flat R P] :
+    Module.Free R P :=
+  free_of_maximalIdeal_rTensor_injective
+    (Module.Flat.rTensor_preserves_injective_linearMap _ Subtype.val_injective)
+
+/--
+If `M → N → P → 0` is a presentation of `P` over a local ring `(R, 𝔪, k)` with
+`M` finite and `N` finite free, then `k ⊗ M → k ⊗ N` is injective implies `P` is free.
+-/
+theorem free_of_lTensor_residueField_injective
+    [Module.Finite R M] [Module.Finite R N] [Module.Free R N]
+    (hf : Function.Injective (f.lTensor k)) :
+    Module.Free R P := by
+  have := Module.finitePresentation_of_free_of_surjective g hg
+    (by rw [h.linearMap_ker_eq, LinearMap.range_eq_map]; exact (Module.Finite.out).map f)
+  apply free_of_maximalIdeal_rTensor_injective
+  rw [← LinearMap.lTensor_inj_iff_rTensor_inj]
+  apply lTensor_injective_of_exact_of_exact_of_rTensor_injective
+    h hg (LinearMap.exact_subtype_mkQ 𝔪) (Submodule.mkQ_surjective _)
+    ((LinearMap.lTensor_inj_iff_rTensor_inj _ _).mp hf)
+    (Module.Flat.lTensor_preserves_injective_linearMap _ Subtype.val_injective)
+
+end Module
+
+/--
+Given a linear map `l : M → N` over a local ring `(R, 𝔪, k)`
+with `M` finite and `N` is finite free,
+`l` is a split injection iff `k ⊗ l` is a (split) injection.
+-/
+theorem LocalRing.split_injective_iff_lTensor_residueField_injective
+    [Module.Finite R M] [Module.Finite R N] [Module.Free R N] (l : M →ₗ[R] N) :
+    (∃ l', l' ∘ₗ l = LinearMap.id) ↔ Function.Injective (l.lTensor (ResidueField R)) := by
+  constructor
+  · intro ⟨l', hl⟩
+    have : l'.lTensor (ResidueField R) ∘ₗ l.lTensor (ResidueField R) = .id := by
+      rw [← LinearMap.lTensor_comp, hl, LinearMap.lTensor_id]
+    exact Function.HasLeftInverse.injective ⟨_, LinearMap.congr_fun this⟩
+  · intro h
+    -- By the lemma above, `k ⊗ l` injective => `N ⧸ l(M)` free.
```

Comment:
It is directly on the next line though.
I've changed the wording a bit but I'll include the lemma name if you still think it's better to do so.

Reply:
I think so, because in the future people might move things around and it won't be "above" anymore (and given the lack of attention that is normally taken at the doc, it is likely they would forget to update).

Reply:
To clarify, the current wording is 
```
    -- By the a previously proved lemma, `k ⊗ l` injective => `N ⧸ l(M)` free.
    have := Module.free_of_lTensor_residueField_injective l (LinearMap.range l).mkQ
```
And you suggest
```
    -- By `Module.free_of_lTensor_residueField_injective`, `k ⊗ l` injective => `N ⧸ l(M)` free.
    have := Module.free_of_lTensor_residueField_injective l (LinearMap.range l).mkQ
```

Reply:
Yes